### ISSUE TITLE
Learning rate scheduler should not need to be told about the number of steps per epoch

### ISSUE THREAD

**Issue by dirkgr** [allenai/allennlp#3707]
Some of the learning rate schedulers take a `num_steps_per_epoch` parameter. It breaks several abstractions to provide it this way, and in the config. At least when we have a non-lazy reader, we should determine that number automatically.

**Comment by brendan-ai2** [comment]
I like this idea. We should also consider doing this in the lazy case as we usually iterate over the data anyway to construct the vocab. For instance, we could wrap the `instances` passed to the `Vocabulary` constructor in an `IterableCounter` and query it after construction.

**Comment by matt-gardner** [comment]
In the lazy case you almost certainly don't want to iterate over the data to construct a vocab, you want to compute the vocab separately.  And I'd be careful about using an `IterableCounter` like that, because some paths through building a `Vocabulary` don't actually result in iterating over the instances.

**Comment by brendan-ai2** [comment]
In that case perhaps `allennlp dry-run`, which generates said vocab, should be augmented to also save information like the total size?

**Comment by matt-gardner** [comment]
I guess we already have `DataIterator.get_num_batches` - can't we just use that?  Getting it to the right place might be tricky, but it's possible we could do something in `TrainModel.from_partial_objects` to grab the right number before passing things on to the learning rate scheduler.

**Comment by matt-gardner** [comment]
Actually doing this right, without causing an iteration over the data in the lazy case unless it's necessary, might be pretty hard, though.

**Comment by dirkgr** [comment]
@DeNeutoy, have you thought about this as part of your re-design of data loading? I think this might fit neatly into the abstraction you proposed (even if many implementations still have to iterate over everything).

**Comment by DeNeutoy** [comment]
I think I missed the discussion about why this is a good idea/a problem... inferring it automatically seems to tie a user to using all of allennlp more, rather than less, if you have to use our dataset/vocab etc to use the learning rate scheduler? What happens in the case that you want the `warmup_steps` to not be a function of the size of your data (if it is very large)? Perhaps it could have a default which is inferred or something. I'm not sure how this would fit in/relate to the new dataloaders, we might need to chat in person.

**Comment by matt-gardner** [comment]
I think the main annoyance is that you have to compute it and put it in your config file.  A relatively minor annoyance, but it is annoying, especially as we should be able to compute this automatically.

The simplest fix I can think of is to put in a bit of logic here: https://github.com/allenai/allennlp/blob/9f238b9996bcb569e58f143f779faab8cd09b971/allennlp/training/trainer.py#L830

We could just add a line like `batches_per_epoch = iterator.get_num_batches(train_data)`, and pass that to the learning rate scheduler as an extra parameter that's available if a scheduler needs it.  Because of how `**kwargs` are handled in `FromParams`, it shouldn't be a problem if a particular scheduler *doesn't* need it.  Seems like a pretty easy fix.  `get_num_batches` already doesn't look at the data if it's lazy and a parameter has been passed to the iterator.

**Comment by dirkgr** [comment]
In the code I have, it very much cares about seeing an extra parameter that it doesn't expect. Do you have an improved version of `FromParams` that's not in master yet?

**Comment by matt-gardner** [comment]
Hmm, can you give a failing test?  If you can make a PR with a minimal failing test, I'll see about fixing it.  I thought what I did in #3705 would make this possible, but there might be something still missing, and I can look into it (not earlier than Wednesday, though).

**Comment by dirkgr** [comment]
I put a PR for this issue at #3718. It will probably not work like this, but it's a start.

**Comment by dirkgr** [comment]
> I thought what I did in #3705 would make this possible.

Hmm, it says "merged 3h ago". I can't be entirely sure I had that change yet.

**Comment by matt-gardner** [comment]
Oh, looks like learning rate schedulers still have a custom `from_params` method, so this almost certainly won't work as-is.  Definitely don't let this block you on what you're doing, but I will make removing that custom `from_params` method my next piece of work, and I'll make sure that it can handle the use that you want when I do it.
